pkg/lib/hooks.js

```diff
@@ -0,0 +1,267 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2020 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import cockpit from 'cockpit';
+import { useState, useEffect, useRef } from 'react';
+import deep_equal from "deep-equal";
+
+/* HOOKS
+ *
+ * These are some custom React hooks for Cockpit specific things.
+ *
+ * Overview:
+ *
+ * - useLocationPath: For following along with cockpit.location.path.
+ *
+ * - useLoggedInUser: For accessing information about the currently
+ * logged in user.
+ *
+ * - useFile: For reading and watching files.
+ *
+ * - useObject: For maintaining arbitrary stateful objects that get
+ * created from the properties of a component.
+ *
+ * - useEvent: For reacting to events emitted by arbitrary objects.
+ *
+ * - useDeepEqualMemo: A utility hook that can help with things that
+ * need deep equal comparisons in places where React only offers
+ * Object identity comparisons, such as with useEffect.
+ */
+
+/* - useLocationPath()
+ *
+ * function Component() {
+ *   const path = useLocationPath();
+ *
+ *   ...
+ * }
+ *
+ * "path" is the current value of cockpit.location.path and the
+ * component is re-rendered when it changes.  "path" is always a valid
+ * array of strings and never null or otherwise uninitialized.
```

Comment:
This raised an eyebrow -- I checked, and it is indeed an array, and the [docs](https://cockpit-project.org/guide/latest/cockpit-location.html#cockpit-location-path`path`) say that as well. `path` is a terrible name for this property: First of all `cockpit.location.path` is an empty array on most pages (e. g. /system/logs) which clearly have a "path". I only saw some entries on the Accounts page when I select a user, so that for `/users#/martin`  the "path" property is `["martin"]`.

Our docs are utterly misleading as well, if that is really meaning to just decode the hash anchor after the path -- how can you ever get multiple of them?

Anyway, that's API, we can't rename it now. But let's fix our documentation.

For this PR, can you please add a link to https://cockpit-project.org/guide/latest/cockpit-location.html#cockpit-location-path ? Thanks!

Reply:
`cockpit.location` is all about navigating inside a page. We encode that in the hash of the global URL. You get multiple entries in the path when there are slashes in the hash, like on the storage page: `/storage#/vg/fedora` gives `[ "vg", "fedora"  ]`.

The path is pre-parsed into an array so that application code doesn't need to worry about encoding things for URLs.

Reply:
Thanks for the info! That makes more sense.

Reply:
Added the link.

Reply:
Do we also need useLocationOptions? Or just `usePageLocation`? Hmm.

Reply:
I have added a PROPOSAL commit for usePageLocation. I like it better. useLocationPath was just the minimum needed for the Accounts page but now I think we should really expose the whole cockpit.location object.